src/module/data.rs

```diff
@@ -72,39 +113,30 @@ impl ModuleData {
         self.arena.iter().map(|(_, f)| f)
     }
 
-    /// Adds a new passive data segment with the specified contents
-    pub fn add(&mut self, value: Vec<u8>) -> DataId {
-        self.arena.alloc_with_id(|id| Data {
-            id,
-            value,
-            passive: true,
-        })
-    }
-
     // Note that this is inaccordance with the upstream bulk memory proposal to
     // WebAssembly and isn't currently part of the WebAssembly standard.
     pub(crate) fn emit_data_count(&self, cx: &mut EmitContext) {
         let mut count = 0;
 
-        // The first elements in the index space will be active segments, so
-        // count all those first. Here we push an "invalid id", or one we know
-        // isn't ever used, as the low data segments.
-        for mem in cx.module.memories.iter() {
-            count += mem.emit_data().count();
-        }
-
-        // After the active data segments, assign indices to the passive data
-        // segments.
-        let mut any_passive = false;
         for data in self.iter() {
             cx.indices.set_data_index(data.id(), count as u32);
             count += 1;
-            any_passive = true;
         }
 
-        if any_passive {
-            cx.start_section(Section::DataCount).encoder.usize(count);
-        }
+        // Technically, we /could/ emit slightly smaller binaries by only adding
+        // this section when either:
+        //
+        // * there are passive data segments, or
+        //
+        // * there are `memory.init` or `data.drop` instructions (99.99% overlap
+        //   with the previous case, but you technically can try and drop active
+        //   data segments and get a runtime error!)
+        //
+        // However, checking that second case can be somewhat expensive (iterate
+        // over every instruction in every function), so we just always emit the
+        // `DataCount` section when we have data segments. At most this is five
+        // unnecessary bytes, so not a big deal in the grand scheme of things.
+        cx.start_section(Section::DataCount).encoder.usize(count);
```

Comment:
While I agree with this comment, this may still be problematic for generating MVP binaries right?

Reply:
Hm, I was assuming that MVP consumers are required to skip sections they don't know about, but maybe not?

Reply:
I don't think so since this is a first-class section rather than a custom section, and you can see some errors in the tests for example

Reply:
Ok, new commit implements this check.

Also fixed the spec-test bugs. Wasn't handling when we *didn't* have the `DataCount` section and didn't pre-allocate data segments.